---
title: "Lab 10: Population genomics based on high throughput sequencing (HTS)"
output:
  rmarkdown::html_document:
    theme: lumen
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part I - Reading VCF data
  Genetic variation data is typically stored in variant call format (VCF) files. One advantage of using VCF files is that only variants (e.g., SNPs, indels, etc.) are reported which economizes files size relative to a format that may included invariant sites.  

## VCF file structure
  A VCF file can be thought of as having three sections: a vcf header (meta region), a fix region and a gt region. The information in the meta region defines the abbreviations used elsewhere in the file. Below the metadata region, the data are tabular. The first eight columns of this table contain information about each variant. The organization of each cell containing a genotype and associated information is specified in column nine, the FORMAT column.

```{r, echo=FALSE, eval=FALSE}
BiocManager::install("vcfR")
```

```{r}
library(vcfR)
```

```{r}
data(vcfR_example)
vcf
```

## The meta region
  The meta region contains information about the file, its creation, as well as information to interpret abbreviations used elsewhere in the file. Each line of the meta region begins with a double pound sign (‘##’).
```{r}
strwrap(vcf@meta[1:7])
```

```{r}
queryMETA(vcf)
```

```{r}
queryMETA(vcf, element = 'DP')
```

```{r}
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```

## The fix region
  The fix region contains information for each variant which is sometimes summarized over all samples. The first eight columns of the fixed region are titled CHROM, POS, ID, REF, ALT, QUAL, FILTER and INFO.
  
```{r}
head(getFIX(vcf))
```

## The gt region
  The gt (genotype) region contains information about each variant for each sample. The values for each variant and each sample are colon delimited.
  
```{r}
vcf@gt[1:6, 1:4]
```

## vcfR
  Read VCF format files into memory using the function read.vcfR()

```{r}
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```


```{r}
head(vcf)
```

```{r}
write.vcf(vcf, "myVCFdata_filtered.vcf.gz")
```


## Excercise Part 1

### 1. How would we find more information about read.vcfR()?
```{r, eval=FALSE}
?read.vcfR
```

### 2. How would we learn what the acronym “AD” stands for?
```{r}
queryMETA(vcf, element = 'AD')
```

### 3. We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?
```{r}
tail(vcf@fix)

## use tail() function to view the last few lines of gt or meta data. 
## head() summarize the information in the all or each of three VCF region.
```

### 4. There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?

```{r}
plot(vcf)
```

```{r}
library(ggplot2)

qplot(getQUAL(vcf), geom = "histogram")
# yes, it is useful to look at the quality of the data being analyzed.
```

### 5. How would we query the sample names?
```{r}
colnames(vcf@gt)
```

# Part II - Analysis of Genome Data

## Opening and examining the dataset
```{r}
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```


```{r}
vcf
```

## Converting VCF data to a genlight object

```{r}
library(adegenet)
```

```{r}
x <- vcfR2genlight(vcf)
```

```{r}
x
```


```{r}
# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]
```


```{r}
# genlight
t(as.matrix(x))[c(1,5,17), 1:3]
```


```{r}
library(adegenet)
library(poppr)
```


```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)
```


```{r}
ploidy(x) <- 2
```


## Distance matrices
```{r}
## To summarize, create a distance matrix from a genlight object using dist()
x.dist <- dist(x)
```

```{r}
x.dist <- poppr::bitwise.dist(x)
```

## Creating chromR objects
```{r}
library(vcfR)
```


```{r}
# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")
```

```{r}
# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)

chrom

plot(chrom)

chromoqc(chrom, dp.alpha = 66)
```

## Processing chromR objects
```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
plot(chrom)
chromoqc(chrom, dp.alpha = 66)
```


```{r}
# vcf <- read.vcfR("pinfsc50_qc.vcf.gz", verbose = FALSE)
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```

```{r}
# Tabular summaries
head(chrom@var.info)

head(chrom@win.info)
```

## Genetic differentiation
```{r}
library(vcfR)
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))

knitr::kable(head(myDiff[,16:19]))
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```


```{r}
## summarize data to use violin plots
library(reshape2)
library(ggplot2)
p <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
p <- ggplot(p, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

## Exercise Part 2

### 1. You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST(y-axis) by genomic position (POS)?

```{r}
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```

### 2. This Manhatttan plot shouldlook a bit unusual. Can you think of anything that may be wrong with this analysis?
```{r}
table(pop)
##it appears that the sample size is small
```

### 3. Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?
```{r}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```


### 4. Can you use the function queryMETA() to look for other data in your file that may be of interest?
```{r}
queryMETA(vcf)
```

