---
title: "Lab8 Sequence Alignment, Phylogenetic Analysis and Tree Visualization"
output:
  rmarkdown::html_document:
    theme: lumen
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# NGPhylogeny,fr
This is a quick and easy working flow including multiple sequence alignment, phylogenetic analysis and tree visualization. To do  phylogenetic analysis involves following steps:
  1. assembling a group of sequences to evaluate
  2. aligning the sequences so that a distance can be calculate between the sequences
  3. determining the relationships among the sequences
  4. visualizing the relationships. 

Once the sequences are aligned, NGPhylogeny.fr is involved with working on steps from step 3, which is analyzing and visulizing the replationship among the sequnces. Here is how you work with NGphylogeny.fr:
  1. Go to the site for [Robust Phylogenetic Analysis For The Non-Specialist](https://ngphylogeny.fr/).
  2. Under Phylogenetic Analysis, select “One Click”.
  3. Upload or paste in your DNA sequencee file.
  4. Click submit. 
  5. FastME gives you output tree in nhx file and you'll all so be able to save in png (or other image format) by using other visualization tools such as iTol. 

### Exercise: Use NGPhylogeny.fr to analysis the set of rRNA sequence provided and put the .png file from your analysis.
```{r}
library(knitr)
library(png)
```

![Tree of life](data/tree.png)


# Phylogenetic Analysis (tree building) on CIPRES
The [CIPRES Science Gateway](https://www.phylo.org/portal2/login!input.action/) is a public resource for inference of large phylogenetic trees. It provides access to large computational resources of the NSF TeraGrid through a simple browser interface. Popular sequence alignment tools such as ClustalW, Muscle and MAFFT and the community tree inference tools FasttreeML, GARLI, RAxML, PAUP, and MrBayes are availabe through CIPRES Science Gateway. Here is breif steps of how to run MAFFT and FasttreeMP:

1. Create a new folder.
2. Upload your .fna file in the Data folder.
3. Slect the input file (For FasttreeMP tool, use the output file obtained from MAFFT) and choose the tool.
4. Set Input Prameters to Advanced Parameters.
5. Save and Run Task.
6. When the task is finished, click on the Label.
7. View on Output and View the output file of your interest (output.mafft or  fastree_result.tre in this case.)
8. Rename and Save to Current Folder. 

TOL_output.mafft file can be uploaded and viewed by MSA visualization application including [Mview](https://www.ebi.ac.uk/Tools/msa/mview/). 

### Exercise: Align and do phylogenetic analysis off the sequences in CIPRES using MAFFT and FastTreeMP.

#### To view the TOL_output.mafft MSA click [here](https://www.ebi.ac.uk/Tools/services/rest/mview/result/mview-I20201124-194908-0211-81171777-p1m/aln-html).



# ggtree tutorial

```{r, echo=FALSE, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")
BiocManager::install("ggimage")
```


```{r}
library(tidyverse)
library(ggtree)
# added treeio and ggimage library
library(treeio)
library(ggimage)
# need to install libmagick 
# sudo apt-get install libmagick++-dev
```

## Basic trees

```{r}
tree <- read.tree("data/tree_newick.nwk")
tree
```

```{r}
# build a ggplot with a geom_tree
ggplot(tree) + geom_tree() + theme_tree()
# This is convenient shorthand
ggtree(tree)
```

```{r}
# add a scale
ggtree(tree) + geom_treescale()
# or add the entire scale to the x axis with theme_tree2()
ggtree(tree) + theme_tree2()
```

```{r}
# disable scaling to produce a cladogram
ggtree(tree, branch.length="none")
```
```{r}
# drawing cladogram (no branch scaling) with aesthetics of the plot
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3)
```

### ggrtree Tutorial Exercise 1

1. Create a slanted phylogenetic tree.
```{r}
ggtree(tree, layout="slanted") + ggtitle("Slanted phylogenetic tree")
```

2. Create a circular phylogenetic tree.
```{r}
ggtree(tree, layout="circular") + ggtitle("Circular phylogenetic tree")
```

3. Create a circular unscaled cladogram with thick red lines.
```{r}
ggtree(tree, layout="circular", branch.length="none", color="red", size=2, linetype=1)
```

### ggrtree Tutorial Exercise 2
Similar to how we change the aesthetics for the tree inside the ggtree() call, we can also change the aesthetics of the points themselves by passing graphical parameters inside the geom_nodepoint() or geom_tippoint() calls. Create a phylogeny with the following aesthetic characteristics:

* tips labeled in purple
* purple-colored diamond-shape tip points (hint: Google search “R point  characters”)
* large semitransparent yellow node points (hint: alpha=)
* Add a title with + ggtitle(...)

```{r}
ggtree(tree) + geom_tiplab(size=4, color="purple") + geom_nodepoint(color="yellow", alpha=1/2, size=3) + geom_tippoint(color="purple", shape=18, size=2)+ ggtitle("Exercise 2 phylogenetic tree")
```

## Tree annotation

```{r}
# Internal node number
ggtree(tree) + geom_text(aes(label=node), hjust=-.3)
```

```{r}
ggtree(tree) + geom_tiplab()

# get the internal node numbers
MRCA(tree, c("C", "E"))
MRCA(tree, c("G", "H"))
```

```{r}
# Labeling clades
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8, align=TRUE) + 
  geom_cladelabel(node=21, label="A different clade", 
                  color="blue", offset=.8, align=TRUE) + 
  theme_tree2() + 
  xlim(0, 70) + 
  theme_tree()
```

```{r}
# Highlight the entire clade
ggtree(tree) + 
  geom_tiplab() + 
  geom_hilight(node=17, fill="gold") + 
  geom_hilight(node=21, fill="purple")
```

```{r}
# Connecting taxa
ggtree(tree) + 
  geom_tiplab() + 
  geom_taxalink("E", "H", color="blue3") +
  geom_taxalink("C", "G", color="orange2", curvature=-.9)
```

### ggrtree Tutorial Exercise 3
Produce the figure below.

1. First, find what the MRCA is for taxa B+C, and taxa L+J. You can do this in one of two ways:
  a. Easiest: use MRCA(tree, tip=c("taxon1", "taxon2")) for B/C and L/J separately.
  b. Alternatively: use ggtree(tree) + geom_text(aes(label=node), hjust=-.3) to see what the node labels are on the plot. You might also add tip labels here too.
2. Draw the tree with ggtree(tree).
3. Add tip labels.
4. Highlight these clades with separate colors.
5. Add a clade label to the larger superclade (node=17) that we saw before that includes A, B, C, D, and E. You’ll probably need an offset to get this looking right.
6. Link taxa C to E, and G to J with a dashed gray line (hint: get the geom working first, then try changing the aesthetics. You’ll need linetype=2 somewhere in the geom_taxalink()).
7. Add a scale bar to the bottom by changing the theme.
8. Add a title.
9. Optionally, go back to the original ggtree(tree, ...) call and change the layout to "circular".

```{r}
MRCA(tree, c("B", "C"))
MRCA(tree, c("L", "J"))
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_hilight(node=19, fill="lavender") + 
  geom_hilight(node=23, fill="pink") + 
  geom_cladelabel(node=17, label="Superclade 17", 
                  color="red2", offset=-7, align=TRUE) + 
  geom_taxalink("C", "E", linetype=2) +
  geom_taxalink("G", "J", linetype=2) + 
  theme_tree2() + 
  ggtitle("Exercise 3 Phylogenetic Tree")
```

```{r}
ggtree(tree, layout = 'circular') + geom_tiplab() + 
geom_hilight(node=19, fill="lavender") + 
geom_hilight(node=23, fill="pink") + 
  geom_cladelabel(node=19, label="Clade 19", 
                  color="blue", offset=.8, align=TRUE) + 
  geom_cladelabel(node=23, label="Clade 23", 
                  color="blue", offset=.8, align=TRUE) + theme_tree2() + 
  ggtitle("Exercise 3 Circular Phylogenetic Tree")
```


## Advanced tree annotation
```{r}
# Read the data
tree <- read.beast("data/flu_tree_beast.tree")
# supply a most recent sampling date so you get the dates
# and add a scale bar
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() 
# Finally, add tip labels and adjust axis
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() + 
  geom_tiplab(align=TRUE, linesize=.5) + 
  xlim(1990, 2020)
```

```{r}
msaplot(p=ggtree(tree), fasta="data/flu_aasequence.fasta", window=c(150, 175))
```

```{r}
msaplot(p=ggtree(tree), fasta="data/flu_aasequence.fasta", window=c(150, 175)) + 
  coord_polar(theta = "y")
```

```{r}
set.seed(42)
trees <- lapply(rep(c(10, 25, 50, 100), 3), rtree)
class(trees) <- "multiPhylo"
ggtree(trees) + facet_wrap(~.id, scale="free", ncol=4) + ggtitle("Many trees. Such phylogenetics. Wow.")
```

## Exercise: Uploade tree file from the FastTreeMP output on CIPRES using treeio.
```{r}
MyTree <- read.tree("data/TOL_fastree_result.tre")

ggtree(MyTree)+
  geom_nodelab() + geom_tiplab(hjust=-.1, size = 3) + xlim(0,2)+
  geom_hilight(node=30, fill="pink", extend = 0.5) +
  geom_hilight(node=25, fill="purple", extend = 0.5)
```

```{r}
MyTree <- read.tree("data/TOL_fastree_result.tre")

ggtree(MyTree) + geom_text(aes(label=node), hjust=-.1)

ggtree(MyTree)+
  geom_nodelab() + geom_tiplab(hjust=-.1, size = 3) + xlim(0,2)+
  geom_hilight(node=30, fill="pink", extend = 0.5) +
  geom_hilight(node=25, fill="purple", extend = 0.5) +
  geom_hilight(node=37, fill="lavender", extend = 0.5) +
  geom_cladelabel(node=30, label="Bacteria", 
                  color="pink", offset=.8, align=TRUE) + 
  geom_cladelabel(node=25, label="Eukarya", 
                  color="purple", offset=.8, align=TRUE) + 
  geom_cladelabel(node=37, label="Archaea", 
                  color="lavender", offset=.8, align=TRUE)
```


    'Archaeoglobus_fulgidus' and 'Korarchaeum_cryptofilum' are also part of the Archaea domain. I was unable to highlight all Archaea domain, since using node 22 for 'geom_hilight' function highlighted all of them.